Unify Releases Across Operating Systems

mre opened the pull request:
This merges the Windows/macOS/Linux build into one job in the release workflow. It follows the same principle as the release build for ripgrep.

Fixes #1875

mre committed:
Unify Releases Across Operating Systems

This merges the Windows/macOS/Linux build into one job in the release workflow.
It follows the same principle as the release build for ripgrep.

Fixes #1875

mre commented:
Will close for now. I don't think there's anything that needs to be done here. As mentioned, we already have a matrix job.

mre committed:
Fix workflow_dispatch get_release error for testing

mre committed:
Remove unused OpenSSL steps

mre committed:
Fix build on Windows due to rlimit unix constraints

mre committed:
Do not use cross for host target x86_64-unknown-linux-gnu

mre committed:
Add comment explaining why cross is disabled for native linux target

mre committed:
Format handle_fd_limits for readability and style

mre committed:
Remove conditional rlimit guard and simplify fd handling

mre committed:
Restore fallback getrlimit logic for Unix

mre committed:
Simplify handle_fd_limits per maintainer feedback

mre commented:
OK, this should be good to go. Thanks, Thomas for the suggestion to run a "test release."

This way, I uncovered an issue with our new file-limits support (https://github.com/lycheeverse/lychee/pull/2106). 

On Windows, I got

```rust
error[E0432]: unresolved imports `rlimit::Resource`, `rlimit::getrlimit`
   --> lychee-bin\src\main.rs:143:18
    |
143 |     use rlimit::{Resource, getrlimit, increase_nofile_limit};
    |                  ^^^^^^^^  ^^^^^^^^^ no `getrlimit` in the root
    |                  |
    |                  no `Resource` in the root
For more information about this error, try `rustc --explain E0432`.
error: could not compile `lychee` (bin "lychee") due to 1 previous error
Caused by:
  process didn't exit successfully:
```

Inside the `rlimit` crate, `increase_nofile_limit(u64::MAX)` evaluates internally via `#[cfg(unix)]` to attempt bumping to the max hard limit. If it fails, it still gracefully returns `Ok(current_soft_limit)`. And on Windows, via `#[cfg(windows)]`, it simply returns `Ok(lim)`. Source code [here](https://github.com/Nugine/rlimit/blob/94c9e171e9d11690756cb15dc7dac07ad9067478/src/tools.rs#L44).

This means `rlimit` is already entirely cross-platform and already provides the *exact fallback semantics* we want internally! The entire `limit_result` match statement with `cfg(unix)` and `cfg(not(unix))` was just redundant boilerplate. :)

This run shows that all binaries build successfully: https://github.com/lycheeverse/lychee/actions/runs/23962659010/job/69895885187

actions/upload-release-asset@v1 is deprecated and relies on older, outdated versions of Node.js. The `gh` CLI tool is natively pre-installed on all GitHub-hosted runners (Ubuntu, macOS, and Windows). This makes it universally available without needing to download external dependencies or rely on third-party actions.

thomas-zahner commented on lychee-bin/src/main.rs:
```diff
@@ -140,15 +140,14 @@ fn read_lines(file: &File) -> Result<Vec<String>> {
     /// Estimate is based on observing `lsof -p $(pgrep lychee)`.
     const BASELINE_OVERHEAD: u64 = 20;
 
-    if let Ok(soft_limit) = increase_nofile_limit(u64::MAX)
```
Oh right, thanks for simplifying this. My wrong assumption was that `getrlimit` is available on Windows. If this would have been the case it would still have lowered lychee's concurrency on Windows if the limit was low. But realising that this is Unix specific it makes sense to simplify it, and do "nothing" on Windows.

mre replied:
Haha, glad we'll soon have the Windows job to automate those checks. 👍

thomas-zahner commented on .github/workflows/release-binary.yml:
```diff
@@ -32,105 +40,188 @@ jobs:
           echo "BIN=$bin" >> $GITHUB_ENV
 
-      - name: Strip binary (cross)
+      - name: Strip release binary (macOS)
```
Why is macOS separate from the rest? Shouldn't we either use the "cross" approach below for all or alternatively the native approach for all as with macOS? (in which case we would remove the docker invocation below)

mre replied:
I can't remember. Think I ran into some problems on macOS, but I'll try again. Might just be that I forgot to merge that part. Thanks for pointing that out.

mre replied:
The separation is intentional: macOS uses the native `strip` on the runner, while Linux cross-compiled targets (arm, aarch64, musl, i686) have to go through the cross Docker image because the host `strip` cannot handle foreign architectures/libc combos reliably.

I also noticed while working on this that the native `x86_64-unknown-linux-gnu` build was not being stripped at all (it skips cross, so the "cross" strip step did not match). Added a dedicated native Linux strip step in 229deb1d.

thomas-zahner commented on .github/workflows/release-binary.yml:
```diff
@@ -32,105 +40,188 @@ jobs:
+          mkdir -p "$ARCHIVE/docs" "$ARCHIVE/complete"
+          cp "$BIN" "$ARCHIVE"/
+          cp README.md "$ARCHIVE"/
+          cp docs/*.md "$ARCHIVE/docs/" || true
```
Why would this fail?

```suggestion
          cp docs/*.md "$ARCHIVE/docs/"
```

mre replied:
Yeah, I added this because `docs` might not contain md files and would return an error in that case. Not sure if we want to make that a hard build error instead, but come to think about it, I think we should.

thomas-zahner commented on .github/workflows/release-binary.yml:
```diff
@@ -140,146 +231,99 @@ jobs:
+
+      - name: Build archive (Unix)
+        shell: bash
+        if: ${{ !startsWith(matrix.os, 'windows') }}
```
The `matrix.os` part confuses me a bit. What does it mean? Maybe worth a comment?

mre replied:
Sure. Basically I want to say "Only run this step on non-Windows builds. I use `startsWith` because GitHub supports different Windows operating systems. Any suggestions for the comment?

thomas-zahner replied:
Oh wait, never mind. I was only confused by the order of the arguments. I would have expected something like `matrix.os.startsWith('windows')` which probably makes no sense. Maybe I've been doing too much Rust recently :sweat_smile: 

It should actually be quite obvious, so no need to comment.

thomas-zahner commented on .github/workflows/release-binary.yml:
```diff
@@ -140,146 +231,99 @@ jobs:
+          echo "ASSET=$ARCHIVE.tar.gz" >> $GITHUB_ENV
+          echo "ASSET_SUM=$ARCHIVE.tar.gz.sha256" >> $GITHUB_ENV
+
+      - name: Build macOS DMG (aarch64 only)
```
What's with non-aarch64 macOS? Does it fall in the Unix category? Then it would be related to my other comment.

mre replied:
Good point. Will check.

mre replied:
Intentional. x86_64 macOS is on its way out and a tar.gz is enough for that target; only Apple Silicon gets a DMG. Added a comment in 229deb1d documenting this and inviting folks to open an issue if they really need an x86_64 DMG.

thomas-zahner commented on .github/workflows/release-binary.yml:
```diff
@@ -140,146 +231,99 @@ jobs:
-
-      - name: Upload .dmg
-        uses: actions/upload-release-asset@v1
+      - name: Upload macOS DMG
```
I don't think this section is necessary. We could just use the same variable names in both cases then we don't need to differentiate in this last step here.

mre replied:
You're right.

mre replied:
Consolidated in 229deb1d: the DMG step now appends to `ASSET`/`ASSET_SUM` and the dedicated `Upload macOS DMG` step is gone. Single upload step handles both.

thomas-zahner reviewed (approved):
Thanks for the PR, looks good! I've left a few comments & questions. Just don't forget to remove the temporary changes you intend to revert before merging.

mre committed:
Remove `

Co-authored-by: Thomas Zahner <thomas.zahner@protonmail.ch>

mre committed:
Address review feedback on release workflow

- Remove test-run scaffolding (tag_name/upload_url fallbacks and
  repo guard on get_release) now that the workflow is verified.
- Add a native strip step for x86_64-unknown-linux-gnu, which does
  not go through cross and was not being stripped.
- Consolidate macOS DMG upload into the single 'Upload release
  archive' step by appending the DMG to ASSET/ASSET_SUM instead of
  using a separate upload step with dedicated variables.
- Document why only aarch64-macos gets a DMG.

mre commented on .github/workflows/release-binary.yml:
```diff
@@ -32,105 +40,188 @@ jobs:
         # Skip step if not in the official repository as it would fail.
         # Instead fail at the end of this workflow.
-        if: ${{ github.repository == 'lycheeverse/lychee' }}
+        if: ${{ github.repository == 'lycheeverse/lychee' && github.event_name == 'release' }}
```
Removed in 229deb1d (the build is verified).

mre commented on .github/workflows/release-binary.yml:
```diff
@@ -14,15 +16,21 @@ on:
-      tag_name: ${{ steps.get_release.outputs.tag_name }}
-      upload_url: ${{ steps.get_release.outputs.upload_url }}
+      tag_name: ${{ steps.get_release.outputs.tag_name || 'test-tag' }}
+      upload_url: ${{ steps.get_release.outputs.upload_url || '' }}
```
Removed in 229deb1d.

mre commented:
Failing test is the flaky WayBackMachine test again...

thomas-zahner commented:
> Failing test is the flaky WayBackMachine test again...

@katrinafyi Addressed this in 434ae96560fd331363a016e9e55ec4df17778822 :)

Thanks again for the PR, I'll test this with the new release

thomas-zahner commented:
Seems to have [worked](https://github.com/lycheeverse/lychee/releases/)! :tada:
Now it's just the [docker step failing](https://github.com/lycheeverse/lychee/actions/runs/24880946622/job/72858998976). It's no news that this is unstable but now it fails even after multiple retries.
I'll investigate this as part of https://github.com/lycheeverse/lychee/issues/2037 and hope to come up with a stable long-term solution. I feel like docker releases were unreliable since a very long time.
